### PR TITLE
HTML-366 Form Schema Preview should show a warning if an answer is not legal for its question

### PR DESCRIPTION
**Pull request by piyush9620**
https://issues.openmrs.org/browse/HTML-366

### CODE REVIEW

**Comment by piyush9620** [comment]
@mogoodrich , i have tried to fix this ticket as follows:
created a list of valid answer ids depending on the question concepts.
created another list of actual answer ids that correspond to the answer concepts ids.
then compared if the actual concept ids are present in the valid concept ids or not.

**Comment by mogoodrich** [comment]
Thanks @piyush9620, I will take a look, hopefully later today... just getting caught up on my Monday morning email now!

File: api/src/main/java/org/openmrs/module/htmlformentry/handler/ObsTagHandler.java [hunk 1]
```diff
@@ -104,6 +113,11 @@ public TagAnalysis validate(Node node) {
             if (concept == null) {
                 analysis.addError(Context.getMessageSourceService().getMessage("htmlformentry.error.invalidConcept",
                     new Object[] { conceptId }, null));
+            } else {
+                Collection<ConceptAnswer> conceptAnswers = concept.getAnswers();
+                for (ConceptAnswer ca : conceptAnswers) {
```
Comment (mogoodrich) [inline api/src/main/java/org/openmrs/module/htmlformentry/handler/ObsTagHandler.java]: I think that you might run into a NPE exception here in some cases... some observations are numeric or free-text (you enter a number or a text value), so they don't have "answeConceptIds"... this should be null-safe, or perhaps you only need to do this if on of the "answer*" node attributes has been specified.

File: api/src/main/java/org/openmrs/module/htmlformentry/handler/ObsTagHandler.java [hunk 2]
```diff
@@ -92,6 +94,13 @@ public TagAnalysis validate(Node node) {
         List<Concept> concepts = new ArrayList<Concept>();
         String conceptId = HtmlFormEntryUtil.getNodeAttribute(node, "conceptId", null);
         String conceptIds = HtmlFormEntryUtil.getNodeAttribute(node, "conceptIds", null);
+        String answerConceptId = HtmlFormEntryUtil.getNodeAttribute(node, "answerConceptId", null);
+        String answerConceptIds = HtmlFormEntryUtil.getNodeAttribute(node, "answerConceptIds", null);
+        String answerClasses = HtmlFormEntryUtil.getNodeAttribute(node, "answerClasses", null);
```
Comment (mogoodrich) [inline api/src/main/java/org/openmrs/module/htmlformentry/handler/ObsTagHandler.java]: Let's not worry about answerClasses for now, it's kind of a separate use case, and it's unlikely that single concept would have *all* elements of a single class in it's answer set.

File: api/src/main/java/org/openmrs/module/htmlformentry/handler/ObsTagHandler.java [hunk 3]
```diff
@@ -92,6 +94,13 @@ public TagAnalysis validate(Node node) {
...
+        String answerConceptSetIds = HtmlFormEntryUtil.getNodeAttribute(node, "answerConceptSetIds", null);
```
Comment (mogoodrich) [inline api/src/main/java/org/openmrs/module/htmlformentry/handler/ObsTagHandler.java]: Let's skip answerConceptIds for now as well and just handle the base use case, when the end use specifies a "conceptId' and then either a "answerConceptId" or an "answerConceptIds".

File: api/src/main/java/org/openmrs/module/htmlformentry/handler/ObsTagHandler.java [hunk 4]
```diff
@@ -112,10 +126,39 @@ public TagAnalysis validate(Node node) {
                 if (c == null) {
                     analysis.addError(Context.getMessageSourceService().getMessage("htmlformentry.error.invalidConcept",
                         new Object[] { conceptIdString }, null));
+                } else {
+                    Collection<ConceptAnswer> conceptAnswers = c.getAnswers();
```
Comment (mogoodrich) [inline api/src/main/java/org/openmrs/module/htmlformentry/handler/ObsTagHandler.java]: In the case of when the user specifics "conceptIds", it is used in conjunction with a single "answerConceptId", and what we actually need to test if that *each* of the concepts referenced has the answerConcept in it's answer set.  You can look at the Obs tag explanation in the HTML Form Entry Reference page for a greater understand of how this works. You can skip this use case now if you like or consider it an "extra credit" part of the ticket, but I don't think this is doing what we want it do here.

File: api/src/main/java/org/openmrs/module/htmlformentry/handler/ObsTagHandler.java [hunk 5]
```diff
@@ -112,10 +126,39 @@ public TagAnalysis validate(Node node) {
...
+                    }
                 }
                 concepts.add(c);
             }
         }
+
+        if(answerConceptId != null) {
+            Concept c = HtmlFormEntryUtil.getConcept(answerConceptId);
+            actualAnswerConceptIDs.add(c.getConceptId());
+        } else if (answerConceptIds != null || answerConceptSetIds != null || answerClasses != null) {
```
Comment (mogoodrich) [inline api/src/main/java/org/openmrs/module/htmlformentry/handler/ObsTagHandler.java]: Per my comments above, let's just handle the answerConceptIds and answerConceptId use cases here, and ignore answerConceptSetIds and answerClasses for now.

**Comment by piyush9620** [comment]
@mogoodrich can u plz check now.
Thanks.

**Comment by piyush9620** [comment]
@mogoodrich  can u plz check now? 
Thanks.

Also about the tests that you asked me to write, i need a few pairs of valid question concept ids along with their valid answer concept ids. can u help me with these?

**Comment by mogoodrich** [comment]
This looks good, thanks @piyush9620 !

For the tests, take a look at the "standardTestDataset.xml" in OpenMRS Core... this dataset is loaded into the in-memory test database before the tests are run.  I haven't checked fully, but I suspect it will have some question and answer concepts you can use.

If this isn't enough, you could potentially add your own test dataset, search for "executeDataSet" for an example of how that is used.

**Comment by piyush9620** [comment]
@mogoodrich, added a few tests and left out others which where redundant.
Thanks.

**Comment by mogoodrich** [comment]
Thanks!  Crazy day for me today, but should be able to review today or tomorrow...

**Comment by mogoodrich** [comment]
Looks good, thanks @piyush9620 !

**Pull request merged**